New Resource: `azurerm_api_management`

torresdal opened the pull request:
Added a new resource for API Management as requested in #1177. Contains the resource, tests and documentation.

ISSUE: It's marked as WIP since there is currently an issue with how the azure-sdk-for-go (or more likely Azure itself) handles deletion of the service when there are associated subnets. Basically the subnets fail to delete because it's still in use by the API Management service, even though it is deleted. I will create a separate issue for this and link back, but any pointers would be appreciated 👍 

Because of the issue mentioned above, the integration test `TestAccAzureRMApiManagement_complete` currently fail.

torresdal commented:
Regarding the issue (#1520) where subnet used by `azurerm_api_management` cannot be deleted, when `azurerm_api_management` is in fact destroyed and should no longer have a dependency to the subnet.

Has this happend with other Azure Terraform resources? It's strange that it takes 30-50 minutes to create the API Mangement resource in Azure, but only seconds to delete it...

torresdal commented:
As requested by @tombuildsstuff I've registered an issue for the subnet issue at https://github.com/Azure/azure-sdk-for-go/issues/2199

tombuildsstuff commented:
@torresdal awesome, thanks - I'm taking a look through this at the moment :)

tombuildsstuff commented:
@torresdal 

> Regarding the issue (#1520) where subnet used by azurerm_api_management cannot be deleted, when azurerm_api_management is in fact destroyed and should no longer have a dependency to the subnet.
> Has this happend with other Azure Terraform resources? It's strange that it takes 30-50 minutes to create the API Mangement resource in Azure, but only seconds to delete it...

We've had this in a few cases, [Virtual Network Gateway's](https://github.com/Azure/azure-rest-api-specs/issues/1233) and [Application Gateway's](https://github.com/Azure/azure-rest-api-specs/issues/2187) - in both of these cases the API affects the entire resource and as such we were blocked from shipping them until they were fixed*. However in this instance this is an optional field, as such I think we should be able to ship the rest of the resource and then add the subnet support back in when the API is fixed; what do you think?

[* Unfortunately we didn't realise the scale of the Application Gateway bug until after we'd shipped it - and now we've shipped it so it's hard to pull from the provider (it's causing us a lot of pain internally, since it's breaking our acceptance tests nightly / leaving test resources kicking around which fill our quota's).]

torresdal commented on azurerm/data_source_api_management.go:
```diff
@@ -0,0 +1,288 @@
+				Computed: true,
+			},
+
+			"sku": apiManagementDataSourceSkuSchema(),
```
Sure, but just to make sure I understand your preference, since I see you have not mentioned `apiManagementDataSourceCertificateInfoSchema`. What I've done is create funcs for schemas that are used multiple locations. Both `apiManagementDataSourceSkuSchema` and `apiManagementDataSourceCertificateInfoSchema` is used twice in the overall data source schema.
  
If this is not a good practice here (or in general 😃 ) I'll inline them.

tombuildsstuff replied:
tbh in this case I think we could probably remove the second SKU in the data source, since we've also removed it from the resource (which is why I originally posted this)

(in general we tend to follow the Go guidelines of duplicating code the first time, and then refactoring it into a method - but it's mostly a gut feeling based on where something's likely to diverge, in this case I'd argue it probably wouldn't since it's the same object in both cases)

torresdal commented on website/docs/r/api_management.html.markdown:
```diff
@@ -0,0 +1,262 @@
+
+Create a Api Management component.
+
+-> **Note:** When `azurerm_api_management` service has dependencies on one or more `azurerm_subnet` resources, referenced using `vnet_subnet_id` or `additional_location.*.vnet_subnet_id` - the `azurerm_subnet`s will fail to be destroyed because of an issue in Azure where a reference to the `azurerm_api_management` service is left behind after its destruction. The `azurerm_api_management` service  ...
```
Should we replace this with an explanation for why Virtual Networks is not supported at the moment?

torresdal commented:
Thanks for pointing out lots of useful improvements in your review @tombuildsstuff! I believe everything should now be up to date.

In addition: 

* I removed everything related to Virtual Networks as you suggested and retested to make sure `destroy` now works - and it does
* Removed `sku` from `additional_location` after verifying that is was not needed as you pointed out

I left you a few comments on a couple of your suggestions (inlining schema function and if we should document why we currently don't support Virtual Networks).

torresdal commented:
Anything else I can do on this before it get merged?

Also, what is your preference regarding changes in master that needs to be resolved in the pull-request branch - merge or rebase?

tombuildsstuff commented:
hey @torresdal 

Sorry for the delay in re-reviewing this - I'd started reviewing this and forgot to hit submit; let me check if the comments are still valid and re-submit this; apologies for the delay here!

> Also, what is your preference regarding changes in master that needs to be resolved in the pull-request branch - merge or rebase?

Rebasing is generally preferred from our side if possible, but we're fairly flexible here :)

Thanks!

tombuildsstuff commented:
@torresdal actually I've just noticed the vendoring here, so I'm going to rebase this & push a commit to resolve that; hope you don't mind!

tombuildsstuff commented on azurerm/resource_arm_api_management.go:
```diff
@@ -0,0 +1,672 @@
+	return nil
+}
+
+func expandAzureRmApiManagementProperties(d *schema.ResourceData, sku *apimanagement.ServiceSkuProperties) *apimanagement.ServiceProperties {
```
can we in-line this within the Create method?

tombuildsstuff commented on azurerm/resource_arm_api_management.go:
```diff
@@ -0,0 +1,672 @@
+		d.Set("management_api_url", props.ManagementAPIURL)
+		d.Set("scm_url", props.ScmURL)
+		d.Set("static_ips", props.StaticIps)
+		d.Set("custom_properties", &props.CustomProperties)
```
rather than setting these directly (which will result in a set error) - could we add a flatten function for this (as below) and convert this list to a `map[string]interface{}` rather than setting `map[string]string` here?

tombuildsstuff commented on azurerm/resource_arm_api_management.go:
```diff
@@ -0,0 +1,672 @@
+		certificates := flattenApiManagementCertificates(d, props.Certificates)
+
+		if err != nil {
+			return err
```
could we wrap this to explain what failed, e.g. 
```
return fmt.Errorf("Error flattening `certificate`: %+v", err)
```

tombuildsstuff commented on azurerm/resource_arm_api_management.go:
```diff
@@ -0,0 +1,672 @@
+		hostnameConfigurations, err := flattenApiManagementHostnameConfigurations(d, props.HostnameConfigurations)
+
+		if err != nil {
+			return err
```
could we wrap this to explain what failed, e.g. 
```
return fmt.Errorf("Error flattening `hostname_configuration `: %+v", err)
```

tombuildsstuff commented on azurerm/resource_arm_api_management.go:
```diff
@@ -0,0 +1,672 @@
+		d.Set("publisher_name", props.PublisherName)
+
+		d.Set("notification_sender_email", props.NotificationSenderEmail)
+		d.Set("created", props.CreatedAtUtc.Format(time.RFC3339))
```
can we remove the creation date, I can't think when this would be likely to be used within Terraform?

tombuildsstuff commented on azurerm/data_source_api_management.go:
```diff
@@ -0,0 +1,403 @@
+
+			"created": {
+				Type:     schema.TypeString,
+				Computed: true,
```
can we remove the creation date, I can't think when this would be likely to be used within Terraform?

torresdal commented:
I believe I've addressed all the requested changes by @tombuildsstuff now (thanks for a thorough review btw!), so hopefully this should be ready to move on.

torresdal commented:
@jeffreyCline Any chance you could you do a final review? Thanks.

katbyte commented on azurerm/resource_arm_api_management.go:
```diff
@@ -0,0 +1,773 @@
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validateApiManagementName,
```
Could we add this to the azure package in a file api_management.go? and then use it for the data source too

katbyte commented on azurerm/resource_arm_api_management.go:
```diff
@@ -0,0 +1,773 @@
+			"publisher_name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validateApiManagementPublisherName,
```
Could we add this to the azure package in a file api_management.go?

katbyte commented on azurerm/resource_arm_api_management.go:
```diff
@@ -0,0 +1,773 @@
+			"publisher_email": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validateApiManagementPublisherEmail,
```
Could we add this to the azure package in a file api_management.go?

katbyte commented on azurerm/resource_arm_api_management.go:
```diff
@@ -0,0 +1,773 @@
+							Type:     schema.TypeInt,
+							Optional: true,
+							Default:  1,
+						},
```
minor, could we validate this? I presume `IntAtLeast(0)` would be a good starting point

katbyte commented on azurerm/resource_arm_api_management.go:
```diff
@@ -0,0 +1,773 @@
+			"security": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Resource{
```
Should this have a `MaxItems: 1` or can there be multiple?

katbyte commented on azurerm/resource_arm_api_management.go:
```diff
@@ -0,0 +1,773 @@
+						"host_name": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
```
A `validation.NoZeroValues` would be nice here

katbyte commented on azurerm/resource_arm_api_management.go:
```diff
@@ -0,0 +1,773 @@
+						"certificate": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
```
A `validation.NoZeroValues` would be nice here

katbyte commented on azurerm/resource_arm_api_management.go:
```diff
@@ -0,0 +1,672 @@
+	var sku *apimanagement.ServiceSkuProperties
+	if skuConfig := d.Get("sku").([]interface{}); skuConfig != nil {
+		sku = expandAzureRmApiManagementSku(skuConfig)
+	}
```
seconded

katbyte commented on azurerm/resource_arm_api_management.go:
```diff
@@ -0,0 +1,773 @@
+	certificates := expandAzureRmApiManagementCertificates(d)
+	hostname_configurations := expandAzureRmApiManagementHostnameConfigurations(d)
+
+	properties := apimanagement.ServiceProperties{
```
We could join the next two object initializations:
```go
apiManagement := apimanagement.ServiceResource{
		Location:          &location,
		ServiceProperties: &apimanagement.ServiceProperties{
             PublisherName:          utils.String(publisher_name),
     		PublisherEmail:         utils.String(publisher_email),
	     	CustomProperties:       custom_properties,
     		AdditionalLocations:    additional_locations,
	     	Certificates:           certificates,
     		HostnameConfigurations: hostname_configurations,
	     },
		Tags:              expandTags(tags),
		Sku:               sku,
	}
```

katbyte commented on azurerm/data_source_api_management_test.go:
```diff
@@ -0,0 +1,53 @@
+	"github.com/hashicorp/terraform/helper/resource"
+)
+
+func TestAccDataSourceAzureRMApiManagement_basic(t *testing.T) {
```
This test is currently failing:

```
azurerm_api_management.test: Error creating API Management Service "acctestAM-1444076724406126485" (Resource Group "amtestRG-1444076724406126485"): apimanagement.ServiceClient#CreateOrUpdate: Failure sending request: StatusCode=0 -- Original Error: autorest/azure: Service returned an error. Status=400 Code="InvalidParameters" Message="Invalid parameter: Setting up Additional Locations is not supported for Sku Type 'Developer'."
FAIL
```

katbyte commented on azurerm/resource_arm_api_management.go:
```diff
@@ -0,0 +1,773 @@
+	}
+
+	err = future.WaitForCompletion(ctx, client.Client)
+	if err != nil {
```
Very minor but this could be:
```go
if err := future.WaitForCompletion(ctx, client.Client); err != nil {
```

katbyte commented on azurerm/resource_arm_api_management.go:
```diff
@@ -0,0 +1,773 @@
+
+	if sku := resp.Sku; sku != nil {
+		if err := d.Set("sku", flattenApiManagementServiceSku(sku)); err != nil {
+			return fmt.Errorf("Error flattening `sku`: %+v", err)
```
I think this should be `Error setting 'sku'`?

katbyte commented on azurerm/resource_arm_api_management.go:
```diff
@@ -0,0 +1,773 @@
+		output["disable_frontend_tls10"] = val
+	}
+
+	if v := input["Microsoft.WindowsAzure.ApiManagement.Gateway.Security.Protocols.Tls11"]; v != nil {
```
Could this changed to use a local helper?

```go
if err := setCustomPropertyFrom(input, "Microsoft.WindowsAzure.ApiManagement.Gateway.Security.Protocols.Tls11", output, "disable_frontend_tls11"); err != nil {
    return nil, err
}

setCustomPropertyFrom(input, path, output, key) {
  if v := input[path]; v != nil {
    val, err := strconv.ParseBool(*v)
    if err != nil {
      return nil, fmt.Errorf("Error parsing `%s` %q: %+v", key, *v, err)
    }
    output[key] = val
  }

  return nil 
}
```

not blocking the pr, just a idea

katbyte commented on azurerm/resource_arm_api_management.go:
```diff
@@ -0,0 +1,773 @@
+	config := configs[0].(map[string]interface{})
+
+	nameConfig := config["name"].(string)
+	name := apimanagement.SkuType(nameConfig)
```
these two lines could be combined:
```go
name := apimanagement.SkuType(config["name"].(string))
```

katbyte commented on azurerm/resource_arm_api_management.go:
```diff
@@ -0,0 +1,773 @@
+			certificate["store_name"] = string(prop.StoreName)
+
+			// certificate password isn't returned, so let's look it up
+			passwKey := fmt.Sprintf("certificate.%d.certificate_password", i)
```
You will have to look these up by name, if someone edits this on the portal the index's may no longer match.

katbyte commented on azurerm/resource_arm_api_management.go:
```diff
@@ -0,0 +1,773 @@
+			}
+
+			// encoded certificate isn't returned, so let's look it up
+			certKey := fmt.Sprintf("certificate.%d.encoded_certificate", i)
```
Same as above comment

katbyte commented on azurerm/resource_arm_api_management.go:
```diff
@@ -0,0 +1,773 @@
+			certKey := fmt.Sprintf("certificate.%d.encoded_certificate", i)
+			if v, ok := d.GetOk(certKey); ok {
+				cert := v.(string)
+				certificate["encoded_certificate"] = cert
```
These could be combined: 
```go
certificate["encoded_certificate"] = v.(string)
```

katbyte commented on azurerm/resource_arm_api_management.go:
```diff
@@ -0,0 +1,773 @@
+		for i, config := range *configs {
+			host_config := make(map[string]interface{}, 0)
+
+			if config.Type != "" {
```
as this is an enum i'm not sure if you need to check for "" ?

katbyte commented on azurerm/resource_arm_api_management.go:
```diff
@@ -0,0 +1,773 @@
+			}
+
+			// certificate password isn't returned, so let's look it up
+			passKey := fmt.Sprintf("hostname_configuration.%d.certificate_password", i)
```
You will have to look these up by name, if someone edits this on the portal the index's may no longer match.

katbyte commented on azurerm/resource_arm_api_management.go:
```diff
@@ -0,0 +1,773 @@
+			}
+
+			// encoded certificate isn't returned, so let's look it up
+			certKey := fmt.Sprintf("hostname_configuration.%d.certificate", i)
```
You will have to look these up by name, if someone edits this on the portal the index's may no longer match.

katbyte commented on azurerm/resource_arm_api_management.go:
```diff
@@ -0,0 +1,773 @@
+		skus = append(skus, sku)
+	}
+
+	return skus
```
I think this could be simplified: `return []interface{}{sku}`

katbyte commented on azurerm/resource_arm_api_management.go:
```diff
@@ -0,0 +1,773 @@
+	return skus
+}
+
+func validateApiManagementName(v interface{}, k string) (ws []string, es []error) {
```
This should be moved to the azure helper in an api_management.go file

katbyte commented on azurerm/resource_arm_api_management.go:
```diff
@@ -0,0 +1,773 @@
+	return
+}
+
+func validateApiManagementPublisherName(v interface{}, k string) (ws []string, es []error) {
```
This should be moved to the azure helper in an api_management.go file

katbyte commented on azurerm/resource_arm_api_management.go:
```diff
@@ -0,0 +1,773 @@
+	return
+}
+
+func validateApiManagementPublisherEmail(v interface{}, k string) (ws []string, es []error) {
```
This should be moved to the azure helper in an api_management.go file

katbyte commented on azurerm/resource_arm_api_management_test.go:
```diff
@@ -0,0 +1,241 @@
+					testCheckAzureRMApiManagementExists("azurerm_api_management.test"),
+				),
+			},
+		},
```
Could we add an import check step here?

katbyte commented on azurerm/resource_arm_api_management_test.go:
```diff
@@ -0,0 +1,241 @@
+				Check: resource.ComposeTestCheckFunc(
+					testCheckAzureRMApiManagementExists("azurerm_api_management.test"),
+				),
+			},
```
Could we add an import check step here?

katbyte commented on website/docs/r/api_management.html.markdown:
```diff
@@ -0,0 +1,223 @@
+  publisher_email               = "company1@terraform.io"
+  notification_sender_email     = "api@terraform.io"
+
+  additional_location {
```
could we take this example and run it through terraform fmt to get the correct indentation?

katbyte commented on website/docs/r/api_management.html.markdown:
```diff
@@ -0,0 +1,223 @@
+
+`sku` block supports the following:
+
+* `name` - (Required) Specifies the plan's pricing tier.
```
Could we list the possible values here?

torresdal commented on azurerm/config.go:
```diff
@@ -442,6 +447,16 @@ func getArmClient(c *authentication.Config) (*ArmClient, error) {
 
+func (c *ArmClient) registerApiManagementServiceClients(endpoint, subscriptionId string, auth autorest.Authorizer, sender autorest.Sender) {
+	ams := apimanagement.NewServiceClientWithBaseURI(endpoint, subscriptionId)
+	setUserAgent(&ams.Client)
```
Not really :-) I followed the same convention as it were for other resources at the time, but I see that has now been refactored to use `configureClient`. So did I.

torresdal commented:
@katbyte I've addressed most of the things you pointed out. What I'm still a bit unsure about, is how to handle sensitive data not returned by the API. As you pointed out in https://github.com/terraform-providers/terraform-provider-azurerm/blob/97f1cfc6c96f60f314d804425ac76a38811d6aec/azurerm/resource_arm_api_management.go#L649 and https://github.com/terraform-providers/terraform-provider-azurerm/blob/97f1cfc6c96f60f314d804425ac76a38811d6aec/azurerm/resource_arm_api_management.go#L721 there is a potential issue with ordering when I extract data from the state using index.

The problem with these two cases is there is nothing that really uniquely identify the certificates. In the first case I can use a combination of `thumbprint` and `store_name`, but that is also problematic since `thumbprint` is Computed. Right now I don't see any other option than to leave it as is. I do hope there is a better solution though...

As for testing these cases, I've used `ImportStateVerifyIgnore` to work around the issue, which you can see here: https://github.com/terraform-providers/terraform-provider-azurerm/blob/97f1cfc6c96f60f314d804425ac76a38811d6aec/azurerm/resource_arm_api_management_test.go#L124 Is this the correct way to handle this?

I've also changed the reference to the API from `github.com/Azure/azure-sdk-for-go/services/apimanagement/mgmt/2017-03-01/apimanagement` to `github.com/Azure/azure-sdk-for-go/services/preview/apimanagement/mgmt/2018-06-01-preview/apimanagemen`, which I posted a qustion about on Slack: https://terraform-azure.slack.com/archives/CB9RVKPDL/p1537003012000200

katbyte commented:
@torresdal,

In the first instance, the list of certs in the resource, as azure is not returning any real way to link what is configured in terraform to azure I am going to suggest we don't even attempt to read it back into state. Drop the thumbprint and have it as a one way 'set and then update when the tf changes' for now.

As for the 2nd situation you can use the `hostname_configuration` to look up the details by looking it up in the old state: `stateOld := d.Get("hostname_configuration").([]interface{})`

And you are correct wrt `ImportStateVerifyIgnore`! that is the exact way to handle this.

katbyte commented:
@torresdal,

If you do no set the property on read, whatever the TF is set to stays in state. So removing 
```
if err := d.Set("certificate", flattenApiManagementCertificates(d, props.Certificates)); err != nil {
			return fmt.Errorf("Error setting `certificate`: %+v", err)
		}
```

means the only time tf will see a change, is if you change the config (i believe, its been some time since i did this)

In the other instance i mean you can iterate across the old one looking for the same hostname:

```
    stateOld := d.Get("hostname_configuration").([]interface{})
    if configs != nil {
		for i, config := range *configs {
            ...
			if config.HostName != nil {
				host_config["host_name"] = *config.HostName

                var old
                for _,o := range stateOld {
                    if o["host_name"] == host_config["host_name] {
                         old = o
                    }
                }

                host_config["certificate_password"] = old["certificate_password"]
			}


```

Does that clarify it?

torresdal commented:
@katbyte I believe I've now fixed everything. 

For case 1, after doing quite a bit of testing, it is my understanding that in order to avoid diffs I need to return the old state during read, so I did this:

```go
// Azure is not returning identifiable certs (no id), so there is no way to link
// what is configured in terraform to Azure. Because of this, we just return the old state.
if err := d.Set("certificate", d.Get("certificate").([]interface{})); err != nil {
	return fmt.Errorf("Error setting `certificate`: %+v", err)
}
```

If I don't, all certificate values will be empty strings, and TF will detect that as a diff.

For case 2 I made a few changes to the schema structure, which I hope will make it better, but in essence the solution is according to your suggestion. Here's an example of the new schema:

```tf
hostname_configurations {
  proxy {
    host_name                    = "api.terraform.io"
    certificate                  = "${base64encode(file("testdata/api_management_api_test.pfx"))}"
    certificate_password         = "terraform"
    default_ssl_binding          = true
    negotiate_client_certificate = false
  }

  proxy {
    host_name                    = "api2.terraform.io"
    certificate                  = "${base64encode(file("testdata/api_management_api2_test.pfx"))}"
    certificate_password         = "terraform"
    negotiate_client_certificate = true
  }

  portal {
    host_name            = "portal.terraform.io"
    certificate          = "${base64encode(file("testdata/api_management_portal_test.pfx"))}"
    certificate_password = "terraform"
  }
}
```

torresdal commented:
@katbyte (cc @tombuildsstuff ) Have you had time to look at this? Eager to get this out in the wild and start adding the other resources...

tombuildsstuff commented on azurerm/helpers/azure/api_management.go:
```diff
@@ -0,0 +1,116 @@
+	return nil
+}
+
+func ResourceArmApiManagementCustomizeDiff(diff *schema.ResourceDiff, v interface{}) error {
```
we actually need to remove this function - so I'm going to resolve this.

tombuildsstuff commented on azurerm/helpers/azure/api_management.go:
```diff
@@ -0,0 +1,116 @@
+	return nil
+}
+
+func validateHostnameConfig(hostnames []interface{}) error {
```
(as above) we're removing this function - so we can remove this

tombuildsstuff commented:
Data Source tests pass:

<img width="484" alt="screenshot 2018-09-28 at 19 15 16" src="https://user-images.githubusercontent.com/666005/46199651-eca87e00-c352-11e8-8d71-34f71333652e.png">

tombuildsstuff commented on azurerm/data_source_api_management.go:
```diff
@@ -0,0 +1,336 @@
+		}
+
+		if prop.PublicIPAddresses != nil {
+			output["static_ips"] = *prop.PublicIPAddresses
```
on reflection, I guess this should probably be `public_ip_addresses` (since `private_ip_addresses` can also be returned for API Management Services on a network)?

torresdal replied:
Renamed to `public_ip_addresses`. Just need to remember to add `private_ip_addresses` when/if we add support for internal vnets. I will retest this (internal vnets) now that we use the preview API, but I expect the same problem to be present.

tombuildsstuff commented on azurerm/resource_arm_api_management.go:
```diff
@@ -0,0 +1,847 @@
+							Computed: true,
+						},
+
+						"static_ips": {
```
(as below) perhaps this should become `public_ip_addresses`?

tombuildsstuff commented on azurerm/data_source_api_management.go:
```diff
@@ -0,0 +1,336 @@
+							Computed: true,
+						},
+
+						"static_ips": {
```
(as below) perhaps this should become `public_ip_addresses` since there can also be Private IP Addresses?

tombuildsstuff commented on website/docs/r/api_management.html.markdown:
```diff
@@ -0,0 +1,223 @@
+
+* `notification_sender_email` - (Optional) Email address from which the notification will be sent.
+
+* `security` - A `security` block as defined below.
```
this wants `(Optional)`

tombuildsstuff replied:
(fixed)

tombuildsstuff commented on website/docs/r/api_management.html.markdown:
```diff
@@ -0,0 +1,223 @@
+
+---
+
+`sku` block supports the following:
```
This is missing an `A ` at the start

tombuildsstuff replied:
(fixed)

tombuildsstuff commented on azurerm/provider.go:
```diff
@@ -78,32 +78,34 @@ func Provider() terraform.ResourceProvider {
+			"azurerm_network_interface":          dataSourceArmNetworkInterface(),
+			"azurerm_network_security_group":     dataSourceArmNetworkSecurityGroup(),
+			"azurerm_notification_hub":           dataSourceNotificationHub(),
+
```
(fixed)

tombuildsstuff commented:
Data Source tests pass:

<img width="470" alt="screenshot 2018-10-01 at 20 34 21" src="https://user-images.githubusercontent.com/666005/46275422-65a90f00-c5b9-11e8-8371-cd8994f8ee90.png">

tombuildsstuff commented:
Resource Tests pass:

<img width="491" alt="screenshot 2018-10-01 at 22 39 08" src="https://user-images.githubusercontent.com/666005/46281316-d278d500-c5ca-11e8-850f-bdefff3993a5.png">

kumlien commented:
Nice work! Is this one already available for us 'normal users'? I dont' see it on  [the terraform site](https://www.terraform.io/docs/providers/azurerm/index.html) and my idea tf-plugin doesn't recognize it... (running  terraform-provider-azurerm_v1.16.0_x4).

torresdal commented:
@kumlien Looks like its scheduled to be in version 1.17.0.
